CRM/Multicompanyaccounting/Hook/Post/ContributionCreation.php

```diff
@@ -6,36 +6,207 @@ class CRM_Multicompanyaccounting_Hook_Post_ContributionCreation {
+   * @return string|null
+   */
+  private function getOwnerOrganizationIdFromContributionFinancialType($financialTypeId) {
+    $incomeAccountRelationId = self::$incomeAccountRelationId;
+    $query = "SELECT fa.contact_id FROM civicrm_entity_financial_account efa
+                      INNER JOIN civicrm_financial_account fa ON efa.financial_account_id = fa.id
+                      WHERE efa.entity_id = {$financialTypeId} AND efa.entity_table = 'civicrm_financial_type' AND efa.account_relationship = {$incomeAccountRelationId}
+                      LIMIT 1";
+    return CRM_Core_DAO::singleValueQuery($query);
+  }
+
+  public function run() {
+    if (!empty($this->ownerOrganizationId)) {
+      $this->updateOwnerOrganization();
+      $this->setInvoiceNumber();
+    }
+    else {
+      // this will terminate the Contribution transaction in CiviCRM core, which will trigger a rollback and prevent the contribution
+      // from getting created.
+      throw new CRM_Core_Exception("Unable to set the owner organisation and the invoice number for the contribution with id: {$this->contributionId}.");
+    }
+  }
+
+  /**
+   * Stores the contribution owner organization.
+   *
+   * @return void
+   */
+  private function updateOwnerOrganization() {
+    ContributionOwnerOrganisation::setOwnerOrganisation($this->contributionId, $this->ownerOrganizationId);
+  }
+
+  /**
+   * Calculates and stores the contribution invoice
+   * number.
+   * The invoice number is calculated as the following:
+   * 1- Using the contribution owner organization, we get
+   * its related Company record, which contains the invoice
+   * prefix and next invoice number. The value is read using
+   * 'SELECT FOR UPDATE' to acquire a row level lock, to prevent
+   * any other contribution from using the same invoice number.
+   * The lock works because CiviCRM starts a transaction while
+   * creating the contribution, then at some transaction it triggers this
+   * hook, then later it commits the transaction. So the queries
+   * here runs as part of the contribution tran transaction.
+   *
+   * 2- Then the prefix is appended to the invoice number, this
+   * will be the contribution invoice number.
+   *
+   * 3- Then the next invoice number is incremented by one
+   * while padding zeros are preserved.
+   *
+   * 4- Finally the contribution invoice_number is set
+   * to the invoice number in from step 2.
+   *
+   * When the controls get back to CiviCRM core,
+   * CiviCRM will commit the transaction, and thus
+   * the lock on the Company row will be released.
+   *
+   * @return void
+   */
+  private function setInvoiceNumber() {
+    $companyRecord = CRM_Core_DAO::executeQuery("SELECT invoice_prefix, next_invoice_number FROM multicompanyaccounting_company WHERE contact_id = {$this->ownerOrganizationId} FOR UPDATE");
+    $companyRecord->fetch();
+
+    $invoiceNumber = $companyRecord->next_invoice_number;
+    if (!empty($companyRecord->invoice_prefix)) {
```

Comment:
@omarabuhussein 
what is the expected behaviour, when `$companyRecord->invoice_prefix` is empty?